src/common/selectors/clusterSelectors.ts

```diff
@@ -66,17 +61,23 @@ export const allSubnetsIPv4 = (
   return !!networks?.every((network) => network.cidr && Address4.isValid(network.cidr));
 };
 
+export const hasIpv6Subnet = (
+  networks: (MachineNetwork | ClusterNetwork | ServiceNetwork)[] | undefined,
+) => {
+  return !!networks?.some((network) => network.cidr && Address6.isValid(network.cidr));
+};
+
 export const allHostSubnetsIPv4 = (subnets: HostSubnets) =>
   subnets.every((subnet) => Address4.isValid(subnet.subnet));
 
-export const isIPv4 = ({
-  machineNetworks,
-  clusterNetworks,
-  serviceNetworks,
-}: Cluster | NetworkConfigurationValues) =>
-  allSubnetsIPv4(machineNetworks) &&
-  allSubnetsIPv4(clusterNetworks) &&
-  allSubnetsIPv4(serviceNetworks);
+export const isDualStack = (cluster: Cluster) =>
```

Comment:
Before,  `isIPv4` returned true if all three elements had only networks using IPv4. So dual stack would be true if ANY of the three elements had at least one item with IPv6.

The new `isDualStack` returns true if ALL three elements have networks using IPv6, so the logic has slightly changed.

Could you please confirm that this is correct? I'm still learning how Dual Stack should work.

Reply:
There should be 2 situations:
1. A cluster could be configured to use DS
2. The cluster is already configured to use DS

For the first case, this happens when the cluster.host_networks include at least one IPv4 and one IPv6 address (excluding loopback addresses probably? @mkowalski WDYT?).
The second occurs when the machine, cluster, and service networks include at least one IPv4 and one IPv6 address

Reply:
The above is correct.

You can assume the hosts *can form* a dual-stack cluster if `cluster.host_networks` have IPv4 *and* IPv6 available there.

The cluster *is already configured* as a dual-stack if all 3 *-networks have IPv4 *and* IPv6 subnets added. The scenario where only some of *-networks have dual-stack subnets but not all of them should be stopped by the validations in the service.

Reply:
I've renamed the functions to `isDualStack` and `canBeDualStack`. I hope that makes things a bit clearer. :slightly_smiling_face: